### PR TITLE
Add required mediation

### LINKED ISSUE #147: Credential request options mediation requirement
The [`CredentialRequestOptions`](https://w3c.github.io/webappsec-credential-management/#dictdef-credentialrequestoptions) dictionary provides [CredentialMediationRequirement](https://w3c.github.io/webappsec-credential-management/#enumdef-credentialmediationrequirement) [mediation](https://w3c.github.io/webappsec-credential-management/#dom-credentialrequestoptions-mediation) that defaults to "optional". 

The other possible values are:

- ["silent"](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-silent),
- ["optional"](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-optional),
- ["conditional"](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional),
- ["required"](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-required)

I think that any call to `.get()` will be [user mediated](https://w3c.github.io/webappsec-credential-management/#user-mediated) (i.e., will show UI). Thus, we should throw a type error for anything but "required" user mitigation.

### PR DESCRIPTION
**Pull request by marcoscaceres**
Closes #147 

The following tasks have been completed:

- [x] [Modified Web platform tests](https://github.com/web-platform-tests/wpt/pull/47704)

Implementation commitment:

- [ ] [WebKit](https://bugs.webkit.org/show_bug.cgi?id=277322)
- [ ] [Chromium](https://g-issues.chromium.org/issues/362340598)
- [ ] Gecko (link to issue)

Documentation and checks

- [ ] Affects privacy
- [ ] Affects security
- [ ] Pinged MDN
- [ ] Updated Explainer

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/digital-credentials/pull/149.html" title="Last updated on Aug 27, 2024, 1:31 AM UTC (2e9b825)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/digital-credentials/149/6192ec4...2e9b825.html" title="Last updated on Aug 27, 2024, 1:31 AM UTC (2e9b825)">Diff</a>

### CODE REVIEW

**Review by timcappalli — APPROVED** [review]

**Review by OR13 — APPROVED** [review]

**Review by bvandersloot-mozilla — COMMENTED** [review]
I think that `"optional"` would also be acceptable here. That would be the page telling the browser that it may show UI, which is all that is being validated, correct?

**Comment by marcoscaceres** [comment]
@bvandersloot-mozilla, right. I guess it depends on how you read the definition of "optional" from Cred Man: 

> **optional**: If credentials can be handed over for a given operation without user mediation, they will be. If [user mediation is required](https://w3c.github.io/webappsec-credential-management/#origin-requires-user-mediation), then the user agent will involve the user in the decision.

My reading from the above, and what I'm worried about, is that leaving it as "optional" invites optionality in presenting UI (which presuppose is always "required"). 

For completeness, here is how required is defined:

> **required**: The user agent will not hand over credentials without [user mediation](https://w3c.github.io/webappsec-credential-management/#user-mediated), even if the [prevent silent access flag](https://w3c.github.io/webappsec-credential-management/#origin-prevent-silent-access-flag) is unset for an origin.

**Comment by marcoscaceres** [comment]
So, I guess we could do things differently... we could state in the spec that "mediation is always required" for this kind of credential, but we need to fix https://github.com/w3c/webappsec-credential-management/issues/248 ... that could allow "optional", but enforce showing UI in prose and in Cred Man.

**Comment by marcoscaceres** [comment]
@samuelgoto, wouldn't mind your thoughts on this.

**Comment by marcoscaceres** [comment]
Ok, so, yeah... per step 8.1 of of Cred Man:

> If options.[mediation](https://w3c.github.io/webappsec-credential-management/#dom-credentialrequestoptions-mediation) is [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional) and interface does not support [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional) [user mediation](https://w3c.github.io/webappsec-credential-management/#user-mediated), return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a "[TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException).

So, we actually just need to state that mediation is required but not as part of the algorithm... will fix.

**Comment by marcoscaceres** [comment]
@TallTed 🦅 👀 🙏? ... 😊

File: index.html [hunk 1]
```diff
@@ -230,6 +230,15 @@ <h2>
...
       [=digital credential=].
     </p>
+    <p>
+      [=Request a credential|Requesting a `DigitalCredential` credential=] does
+      not support {{CredentialMediationRequirement/"conditional"}} [=user
+      mediation=]. [=User mediation=] is always
+      {{CredentialMediationRequirement/"required"}}. No other
+      {{CredentialMediationRequirement}} enum value is supported, apart from
+      {{CredentialMediationRequirement/"required"}}, and results in a
+      {{TypeError}} when {{CredentialsContainer/get()}} is called.
```
Comment (TallTed) [inline index.html:240]:
```suggestion
      {{CredentialMediationRequirement/"required"}}. The only supported
      {{CredentialMediationRequirement}} enum value is
      {{CredentialMediationRequirement/"required"}}, which results in a
      {{TypeError}} when {{CredentialsContainer/get()}} is called.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Small, human-facing.

**Review by samuelgoto — APPROVED** [review]
LGTM++

> I think that "optional" would also be acceptable here. That would be the page telling the browser that it may show UI, which is all that is being validated, correct?

I share the intuition that we may want to support `optional` too at some point, but I think we can cross the bridge of `optional` when we get there without cornering ourselves.

**Comment by marcoscaceres** [comment]
For "required" and "optional"... I think I have a better solution for all this:
https://github.com/w3c/webappsec-credential-management/issues/256

We need fix the Cred Man API to allow us to set our own default.

**Comment by marcoscaceres** [comment]
@samuelgoto, @bvandersloot-mozilla, here's a first stab at fixing this in Cred Man:
https://github.com/w3c/webappsec-credential-management/pull/258

**Comment by marcoscaceres** [comment]
For those watching at home, this would mean that the following would not throw:

```JS
navigator.identity.get({digital: ...}); // this throws today, which is annoying. 
navigator.identity.get({digital: ..., mediation: "required"});
```

**Comment by marcoscaceres** [comment]
(also, `"optional"` mentioned by a few folks above is meant to be `"conditional"`... yay, naming things if fun 😜)

**Comment by samuelgoto** [comment]
> (also, "optional" mentioned by a few folks above is meant to be "conditional"... yay, naming things if fun 😜)

I'm not sure I follow this reference, but I was actually discussion `optional` rather than `conditional` as:

https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-optional

**Comment by marcoscaceres** [comment]
Sorry @samuelgoto, I might have confused myself 🙈

File: index.html [hunk 2]
```diff
@@ -230,6 +230,15 @@ <h2>
       The <dfn>DigitalCredential</dfn> interface represents a conceptual
       [=digital credential=].
     </p>
+    <p>
+      [=Request a credential|Requesting a `DigitalCredential` credential=] does
+      not support {{CredentialMediationRequirement/"conditional"}} [=user
```
Comment (samuelgoto) [inline index.html:235]:
```suggestion
      not support {{CredentialMediationRequirement/"conditional"}} and {{CredentialMediationRequirement/"optional"}} and {{CredentialMediationRequirement/"silent"}} [=user
```

File: index.html [hunk 3]
```diff
@@ -230,6 +230,15 @@ <h2>
...
+      mediation=]. [=User mediation=] is always
+      {{CredentialMediationRequirement/"required"}}. The only supported
+      {{CredentialMediationRequirement}} enum value is
+      {{CredentialMediationRequirement/"required"}}, which results in a
+      {{TypeError}} when {{CredentialsContainer/get()}} is called.
```
Comment (samuelgoto) [inline index.html:240]:
```suggestion
      {{TypeError}} when {{CredentialsContainer/get()}} is called with anything but {{CredentialMediationRequirement/"required"}}.
```

**Review by samuelgoto — APPROVED** [review]
LGTM++

Just a few minor clarifications.

**Comment by samuelgoto** [comment]
(added chromium bug for implementation)

File: index.html [hunk 4]
```diff
@@ -230,6 +230,15 @@ <h2>
...
     </p>
+    <p>
+      [=Request a credential|Requesting a `DigitalCredential` credential=] does
+      not support {{CredentialMediationRequirement/"conditional"}} and {{CredentialMediationRequirement/"optional"}} and {{CredentialMediationRequirement/"silent"}} [=user
+      mediation=]. [=User mediation=] is always
+      {{CredentialMediationRequirement/"required"}}. The only supported
+      {{CredentialMediationRequirement}} enum value is
+      {{CredentialMediationRequirement/"required"}}, which results in a
+      {{TypeError}} when {{CredentialsContainer/get()}} is called with anything but {{CredentialMediationRequirement/"required"}}.
+    </p>
```
Comment (marcoscaceres) [inline index.html:241]:
```suggestion
    [=User mediation=] is always {{CredentialMediationRequirement/"required"}}.
    [=Request a credential|Requesting a DigitalCredential credential=] does not support {{CredentialMediationRequirement/"conditional"}}, {{CredentialMediationRequirement/"optional"}}, or {{CredentialMediationRequirement/"silent"}} [=user mediation=]. If {{CredentialsContainer/get()}} is called with anything other than {{CredentialMediationRequirement/"required"}}, a {{TypeError}} will be thrown.
    </p>
```

**Comment by samuelgoto** [comment]
As we were implementing this PR, something occurred to us:

(a) first, this is a backwards incompatible change for us: `mediation` is an optional field and it defaults to `optional`, so some developers have been using the default value.
(b) second, is excluding `optional` really required? meaning, isn't the website leaving the choice to the user agent when they choose `optional`? meaning, wouldn't a valid implementation of `optional` be to threat it as `required`?

If this is correct, then can we make this more flexible and also allow the default value to be used, `optional` (and allow each implementation to implement it in whichever way it see fit)?

**Comment by samuelgoto** [comment]
> As we were implementing this PR here, something occurred to me while reviewing it:

FWIW, we found a resolution to this here:

https://github.com/WICG/digital-credentials/issues/175#issuecomment-2397438963

**Pull request merged**